micropython/lora: Add async-capable LoRa modem driver

projectgus opened the pull request:
Working LoRa drivers for:

* SX1276, SX1277, SX1278, SX1279 modems
* SX1261, SX1262 modems.

Eventual goal (not in this PR) is to have enough support for MicroPython to be a LoRaWAN Class A device with a fully async API available.

## How to Test this PR

The [lora README added in the PR branch](https://github.com/projectgus/micropython-lib/blob/feature/lora/micropython/lora/README.md) has instructions for how to use the LoRa drivers.

However, until the PR is merged you need some additional steps to install the lora MicroPython packages.

### Installing with mip

The `--index` argument of mip install is [needed to install packages from a PR branch](https://github.com/micropython/micropython-lib/#installing-packages-from-forks).

Example:

```sh
$ mpremote connect /dev/ttyUSB0 mip install --index https://projectgus.github.io/micropython-lib/mip/feature/lora lora-sx127x lora-sync
```

(Replace `/dev/ttyUSB0` with the serial port you are using, and replace `lora-sx127x` and `lora-sync` with the appropriate packages for your LoRa hardware and programming style  - see the [lora README]((https://github.com/projectgus/micropython-lib/blob/feature/lora/micropython/lora/README.md)) for explanation.)

Or from a REPL:

```python
import mip
mip.install("lora-sync", index="https://projectgus.github.io/micropython-lib/mip/feature/lora")
```

(Replace the package name as needed.)

### Freezing lora into the firmware

If your board doesn't support mip, you can freeze the lora modules into the firmware.

* First, you need to be able to build the MicroPython firmware for your board locally, with the micropython source cloned from the github repo.
* Next, switch the branch of the `micropython-lib` git submodule, as found in the `lib/micropython-lib` directory of the micropython source, to the branch for this PR. [GitHub CLI client](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally?tool=cli) is the easiest way to check out a PR branch from the command line: Running commands similar to `cd lib/micropython; gh pr checkout 501` should do it.
* Update the board [manifest file](https://docs.micropython.org/en/latest/reference/manifest.html) to add the lora packages you need.
* Rebuild the firmware.

## TODOs

* [x] Examples are currently hardcoded to the LoRa modems I was using for testing. I think the best solution here is to just make them say `modem =   # TODO put your modem config here`, but haven't made this change yet.
* [x] Confirm the API & AsyncModem wrapper design is a reasonable approach
* [x] Support ThreadSafeFlag.clear() in MP
* [x] Figure out why the RxTimeout interrupt line is toggling during TX (currently using `_irq_en` as a workaround)
* [x] Split out the single configure() function to allow fast runtime configuration of parameters changed on the fly by the LoRAWAN MAC layer. (Done by having configure() keep the current value for any missing keys, instead of setting defaults.)
* [x] Decide about support for interleaving rx,tx as managed by the driver - i.e. it would be nice if one task could be continuously (or arbitrarily) calling `receive()` while another task sometimes called `transmit()`, and have the modem driver handle the state changes internally without the caller needing to do anything. This isn't necessary for LoRAWAN, but it would make some other types of application logic much simpler. (Implemented.)
* [x] Decide about support for `rx_continuous` mode. If the previous dot point can be implemented then it might not be needed to explicitly support putting the modem into continuous RX mode. (Implemented.)
* [x] Review use of `assert` in the code - some places are probably ValueErrors, some places may be unnecessary.
* [x] Add support for a second modem apart from SX1276. (Done, SX1262.)
* [x] Resolve issue where SX1276 invert_iq_tx setting seems to be inverted from all other modems (this may be a hardware compatibility issue, but it's unusual!)
* [x] Check TODOs in the code, mostly these are things that could be configurable but currently aren't.
* [x] Proper packaging layout
* [x] Documentation

## Follow-up work

Intended for future PRs:

* :heavy_check_mark:  ~~Document the `ant_sw` feature that allows callbacks for switching off-modem antenna switches for RX, TX.~~
* Document `calibrate()` and `calibrate_image()` functions.
* :heavy_check_mark: ~~Document the modem `crc_errors` counter field, and maybe break out CRC checking to its own doc section.~~
* Add some packages for common modules and boards (some have fixed pins, others have specific antenna switch configs). This will make it a lot easier to get up and running with a LoRa development board or module.
* :heavy_check_mark: ~~SX126x radios with DIO2 configured as "enable TCXO" line currently take around 18ms to start the TCXO, which can be made configurable.~~ We can also add a configuration option to not automatically stop the TCXO (and go to "RC Standby") when a receive or transmit ends, which allows tighter receive/transmit timing.
* STM32WL5x is a special case of the SX126x which requires hardware support in MicroPython, have moved to #572.
* OCP limit is not currently configured either automatically or manually by the driver, and possibly should be.
* Add a function to read the instantaneous RSSI at any time during receive.
* Add a documentation section about common hardware issues - clock drift, symptoms of incorrect antenna pin setting, known modem bugs, etc.

## Example API usage

This PR adds two examples, each with synchronous and async API versions:

* `simple_rxtx` example that just transmits and receives in a loop
* `reliable_delivery` which implements a relatively simple "ACK or retry" scheme to receive packets reliably from low power nodes.

This work was funded through GitHub Sponsors.

beyonlo commented:
Hello @projectgus 

Great. This is a great lib for the MicroPython system! 

1. This lib support all features of the SX1276?
2. This lib is really specific to the SX1276 or support other models?

Thank you very much.

beyonlo commented:
Sorry, I see above now that is specific for the SX1276.

beyonlo commented:
I was questioning that because I would like to use in the feature the 2.4Ghz LORA https://www.semtech.com/products/wireless-rf/lora-24ghz for indoor and outdoor location - It has ~3 meters accuracy, like as the [SX1280](https://www.semtech.com/products/wireless-rf/lora-24ghz/sx1280)

projectgus commented:
@beyonlo The idea is to have a generic LoRA modem API that can be extended to other modem models in the future, such as the SX1280. However at least for now the focus is on messaging and supporting LoRAWAN nodes, rather than ToF range measurement.

beyonlo commented:
@projectgus No problem. To have a Micropython driver for  messaging on LORA nodes, will be great :)

dpgeorge commented:
I've had a play with this driver using 2x SX1276 modules connected to a PYBLITEv1.0 and a PYBD-SF2.  Then they run the `rxtx_sync.py` demo code.  I didn't wire up the D0/D1 or RST pins, so it's just using SPI and CS.

It works!  Both devices can send and receive to/from each other.

For the power consumption, it seems quite high.  I measure about 13mA when in RX mode (on top of the pyboard current), and about 30-45mA when transmitting a message (again, on top of pyboard current).  Is that expected?

dpgeorge commented:
> Decide about support for interleaving rx,tx as managed by the driver - i.e. it would be nice if one task could be continuously (or arbitrarily) calling `receive()` while another task sometimes called `transmit()`, and have the modem driver handle the state changes internally without the caller needing to do anything.

I think this would be very useful, for the driver to automatically manage switching between TX and RX.

For LoRaWAN, or for a custom LoRa protocol, is the idea to have RX constantly on (except when transmitting), or for RX only to be active periodically to reduce power consumption?

dpgeorge commented on micropython/lora/sx1276.py:
```diff
@@ -0,0 +1,620 @@
+        self._nss = nss
+
+        self._dio0 = Pin(dio0, mode=Pin.IN) if (dio0 is not None) else None
+        self._dio1 = Pin(dio1, mode=Pin.IN) if (dio1 is not None) else None
```
The user should pass in `Pin` objects themselves, rather than pin id's.  That way, this driver doesn't need to access the SPI and Pin classes at all.

(And the user can still pass in `None` here if they don't want to use these pins.)

dpgeorge commented on micropython/lora/sx1276.py:
```diff
@@ -0,0 +1,620 @@
+
+        if radio_reset:
+            # If the user supplies a reset pin argument, reset the radio
+            radio_reset = Pin(radio_reset, mode=Pin.OPEN_DRAIN, value=1)
```
As above, the user should pass in a `Pin` object, not an id.

dpgeorge commented on micropython/lora/sx1276.py:
```diff
@@ -0,0 +1,620 @@
+
+
+class SX1276:
+    def __init__(self, spi, nss, dio0=None, dio1=None, radio_reset=None, lora_cfg=None):
```
IMO `nss` should be called `cs` to be consistent with "chip select" used in other drivers.

dpgeorge commented on micropython/lora/sx1276.py:
```diff
@@ -0,0 +1,620 @@
+            self._spi.readinto(buf)
+            if __DEBUG__:
+                print("R {:#x} <== {}".format(reg, binascii.hexlify(buf)))
+        finally:
```
Does it need a try-finally?  I don't think SPI read functions can raise (??).

projectgus replied:
Good point, will only fail of `buf` or `reg` are invalid and these are supplied by the object itself.

dpgeorge commented on micropython/lora/sx1276.py:
```diff
@@ -0,0 +1,620 @@
+    def _reg_readinto(self, reg, buf):
+        try:
+            self._nss(0)
+            self._spi.read(1, reg)
```
To eliminate a memory allocation with this read, you can do:
```py
self._spi.readinto(self._buf1, reg)
```
where `self._buf1 = bytearray(1)` is in the constructor.

dpgeorge commented on micropython/lora/sx1276.py:
```diff
@@ -0,0 +1,620 @@
+    def _reg_read(self, reg):
+        buf = bytearray(1)
+        self._reg_readinto(reg, buf)
+        return buf[0]
```
This function can be made more efficient, eg:
```py
self._buf2[0] = reg
self._cs(0)
self._spi.write_readinto(self._buf2, self._buf2)
self._cs(1)
return self._buf2[1]
```

dpgeorge commented on micropython/lora/sx1276.py:
```diff
@@ -0,0 +1,620 @@
+        will_irq = self.start_receive(timeout_ms, rx_continuous)
+        while True:
+            if will_irq:
+                machine.idle()
```
IMO this `idle()` should always be called, even if `will_irq` is False.  Because otherwise this is a very busy loop that consumes lots of power on the MCU side.

projectgus replied:
@dpgeorge I guess this explains the high power consumption. I was worried on some ports this might get stuck in a WFI or similar if no interrupt was set, but I suppose there's always a timer tick or something to wake the CPU up again?

projectgus replied:
Was reminded the original version of this code called `machine.lightsleep(10)` here. Maybe:

```python
if will_irq:
   machine.idle()
else:
   machine.lightsleep(1)
```

is a reasonable compromise? Max 1ms of lag on receive shouldn't be a problem even for a Class B LoRaWAN "Beacon" devices, I don't think.

dpgeorge replied:
`machine.idle()` is perhaps a bit under-specified in the docs... we don't really have any tickless ports (*) so `idle()` should always wake after some short period.  Thus it would be safe to use it unconditionally here.

If you're concerned, then `time.sleep_ms(1)` would be an alternative.  Don't use `machine.lightsleep()`!  That will put it in a very low power consumption mode that usually disables USB.

(*) zephyr port is perhaps tickless, and `machine.idle()` there calls `k_yield()`.  I'm not sure exactly how that works if there are no interrupts to wake it up.

dpgeorge commented on micropython/lora/async_modem.py:
```diff
@@ -0,0 +1,47 @@
+import uasyncio
+
+
+def _clear_irqflag(thread_safe_flag):  # TODO: implement ThreadSafeFlag.clear() in MP
```
`ThreadSafeFlag.clear()` is now implemented.

projectgus commented:
> For the power consumption, it seems quite high. I measure about 13mA when in RX mode (on top of the pyboard current), and about 30-45mA when transmitting a message (again, on top of pyboard current). Is that expected?

That sounded high to me also, but SX1276 datasheet says "Supply current in receive mode" between 10.8 and 11.5mA for AU915 depending on which antenna is used. Transmit current 20-120mA depending on transmit power level.

So probably to be expected.

> >    Decide about support for interleaving rx,tx as managed by the driver - i.e. it would be nice if one task could be continuously (or arbitrarily) calling receive() while another task sometimes called transmit(), and have the modem driver handle the state changes internally without the caller needing to do anything.
>  For LoRaWAN, or for a custom LoRa protocol, is the idea to have RX constantly on (except when transmitting), or for RX only to be active periodically to reduce power consumption?

LoRaWAN specifies particular time slices when the node has to receive. 

Given how much of a power drain constant receiving is, it won't be desirable for battery-powered devices. It's likely most custom LoRa setups will have at least one end on battery power, so it's probably not that useful to have seamless RX/TX switching in the driver. Will pull it out.

dpgeorge commented:
> So probably to be expected.

Yeah, I came to the same conclusion reading the SX datasheet.

> Will pull it out.

OK.  We can always put it back in when we have more experience with this driver.  For now, best to keep it simple/minimal.

projectgus commented:
@dpgeorge Thanks for the review, I've updated based on your comments and did some additional memory usage optimisations (especially in sx1262.py).

The stm32wl5 support has now been split into an sx1262.py driver, tested with a local module, and I've moved the small amount of additional changes needed for the stm32wl5 into another branch for a separate PR. This means this PR no longer requires any MicroPython changes, and @jimmo your modules should now work properly with it!

dpgeorge commented:
Looking good!

One thing that's needed is a `manifest.py` for packaging.  Also to consider how to organise the files in this package, and whether they should be split out into separate packages or components within the one package.  For example, as a user I probably only want either the SX1276 or SX1262 driver, but not both.  And I probably only want to use AsyncModem or Modem, but not both.

Probably the directory `micropython/lora` should have subdirs in it for the actual packages, like `micropython/lora/sx1276` and `micropython/lora/lora-modem` etc.   See `aioble` for a good example of how to structure things.  @jimmo will probably have opinions here :smile:

projectgus commented:
@dpgeorge Can do! Sorry, I did know I needed to do this but I hadn't put it on the TODO list. Will crib from `aioble` and come up with something soon.

`Modem` is actually an abstract base class, so everyone will need that - I will rename it to BaseModem or something and make sure everything is laid out to make this clearer too.

tve commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+        )
+        old_value &= _OPMODE_MODE_MASK
+        if old_value == _OPMODE_MODE_SLEEP:
+            time.sleep_ms(1)  # waking from sleep takes TS_OSC = 250us (typical)
```
why is this delay needed?

projectgus replied:
I think this can be sleep_us(250), will change. Somehow I missed that API call was available.

(My understanding of TS_OSC is that the modem won't accept any further SPI commands from the host for minimum of this long after coming out of sleep mode. Is that wrong?)

tve replied:
Where did you see that in the datasheet? I don't think I have used sleep mode on the sx1276, so I don't have any off-datasheet experience.

projectgus replied:
You're right, I don't know why I thought this was necessary. Looks like it's only necessary for FIFO access over SPI, and I don't think this will happen within TS_OSC after coming out of sleep. Will take it out.

tve commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+        if _DEBUG:
+            print("ISR {}".format(pin))
+        if self._irq_callback:
+            self._irq_callback()
```
How is the callback supposed to distinguish successful TX, successful RX, RX with CRC error, and RX timeout? Why not pass it that info, which the driver already has?

projectgus replied:
The purpose here is to wake up the code blocked in `receive()` or `transmit(), which then reads the interrupt status (and it already knows this information, as you note.)

projectgus replied:
The IRQ callback still doesn't get any modem state, but I still think this is OK. What the implementer is supposed to do is now different and I hope clearer, and also documented.

tve commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+            self._irq_callback()
+
+    def start_receive(self, timeout_ms=None):
+        mode = self._get_mode()
```
unused?

tve commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+            _IRQ_RX_DONE | _IRQ_RX_TIMEOUT,
+        )  # clear flags
+        if flags & _IRQ_RX_TIMEOUT:  # Modem timed out
+            return None
```
This code assumes that end_receive is being called because receive-complete got signalled. What if the upper layer just wants to TX a packet and thus calls end_receive to stop the current RX mode? As far as I can tell, it's incorrect to just call start_transmit...

projectgus replied:
There is now a "low level API" with documentation and (I hope) less arbitrary behaviour.

tve reviewed (commented):
Nice code overall!

I'm having a hard time picturing how to use this driver, there are a bunch of things missing and few that make usage difficult. If this driver is supposed to be a low level driver that can be used for different high-level LoRa uses then I would expect it to look somewhat different:
- break all sub-parts of `config()` out into separate functions so they can be called when needed, examples are changing BW/SF and TX power in response to ACKs (or lack thereof) or changing channels (freq) (see https://github.com/lemariva/uPyLoRaWAN/blob/master/sx127x.py for example where the init is a collection of calls to these broken out functions)
- ability to have continuous RX as the default mode as opposed to switching back to standby (that's what I tend to use for powered devices)
- provide functionality to check for RX in-progress before TX
- provide a function to calculate packet duration (with LoRa it's not that trivial and often needed since air time can be quite long)
- provide timing information, in particular when a packet was received (I find this essential to sync nodes e.g. to define an ACK window when there is no other shared clock, such as GPS)
- provide functions to enter and leave sleep mode, incl. setting all pins to high impedance

I find the way the IRQ flags are managed rather fragile. All the masking/unmasking only works if start_xx/end_xx are always correctly called in the right sequence. If they're not, one is guaranteed odd difficult to troubleshoot results. I'm more used to clear _all_ flags _before_ switching to TX/RX mode so I'm guaranteed I'm starting from a known state for that phase. Also, I've never used the IRQ masks on the semtech chips and rather do a dispatch in the IRQ handler to figure out what happened. You seem to experience some spurious interrupt according tot he comments and I wouldn't be surprised if it had to do with changing mode, changing IRQ masks, and changing DIO Mapping in back-to-back SPI transactions. There's a wimpy little uC in that sx1276 that has to keep up and do the right thing kepping in mind that the mode change may take way longer than the subsequent SPI transactions (we know that little uC has bugs...)

WRT polling for RX I would not provide any examples or high level code that polls the radio for RX complete. It's a noise disaster in my experience. (Maybe different in stm32wle.) I believe the only appropriate non-interrupt usage pattern for RX is to have an RX window, start the RX at the beginning of the window, wait, and then check whether something arrived _once_ at the end of the window.

I did not see high-level code to operate the driver in always-RX mode and I believe it's rather tricky. Specifically because the IRQ handler gets no info about why it got called and there's an issue with calling RX when no RX-complete indication has been provided (e.g. because one wants to TX).

Documentation???

tve commented on micropython/lora/lora/async_modem.py:
```diff
@@ -0,0 +1,42 @@
+        self._modem.end_transmit()
+
+    def _callback(self):
+        self._flag.set()
```
What happens if I first call `receive` and then while it's waiting I call `transmit`?

projectgus replied:
Yes, this is not at all supported right now. Need to either make it clear that the same task should be calling both receive and transmit, or re-add support for cancelling receive.

projectgus replied:
Now supported!

tve reviewed (commented):
Nice code overall!

I'm having a hard time picturing how to use this driver, there are a bunch of things missing and few that make usage difficult. If this driver is supposed to be a low level driver that can be used for different high-level LoRa uses then I would expect it to look somewhat different:
- break all sub-parts of `config()` out into separate functions so they can be called when needed, examples are changing BW/SF and TX power in response to ACKs (or lack thereof) or changing channels (freq) (see https://github.com/lemariva/uPyLoRaWAN/blob/master/sx127x.py for example where the init is a collection of calls to these broken out functions)
- ability to have continuous RX as the default mode as opposed to switching back to standby (that's what I tend to use for powered devices)
- provide functionality to check for RX in-progress before TX
- provide a function to calculate packet duration (with LoRa it's not that trivial and often needed since air time can be quite long)
- provide timing information, in particular when a packet was received (I find this essential to sync nodes e.g. to define an ACK window when there is no other shared clock, such as GPS)
- provide functions to enter and leave sleep mode, incl. setting all pins to high impedance

I find the way the IRQ flags are managed rather fragile. All the masking/unmasking only works if start_xx/end_xx are always correctly called in the right sequence. If they're not, one is guaranteed odd difficult to troubleshoot results. I'm more used to clear _all_ flags _before_ switching to TX/RX mode so I'm guaranteed I'm starting from a known state for that phase. Also, I've never used the IRQ masks on the semtech chips and rather do a dispatch in the IRQ handler to figure out what happened. You seem to experience some spurious interrupt according tot he comments and I wouldn't be surprised if it had to do with changing mode, changing IRQ masks, and changing DIO Mapping in back-to-back SPI transactions. There's a wimpy little uC in that sx1276 that has to keep up and do the right thing kepping in mind that the mode change may take way longer than the subsequent SPI transactions (we know that little uC has bugs...)

WRT polling for RX I would not provide any examples or high level code that polls the radio for RX complete. It's a noise disaster in my experience. (Maybe different in stm32wle.) I believe the only appropriate non-interrupt usage pattern for RX is to have an RX window, start the RX at the beginning of the window, wait, and then check whether something arrived _once_ at the end of the window.

I did not see high-level code to operate the driver in always-RX mode and I believe it's rather tricky. Specifically because the IRQ handler gets no info about why it got called and there's an issue with calling RX when no RX-complete indication has been provided (e.g. because one wants to TX).

Documentation???

projectgus commented:
Hey @tve,

Thanks for taking a detailed look through this. I appreciate that you seem to be pretty familiar with these radios, it's great to get that perspective as I'm no LoRa expert at this stage.

May defer a few of the in-code replies until I have a chance to check properly on the answers.

>     break all sub-parts of config() out into separate functions so they can be called when needed, examples are changing BW/SF and TX power in response to ACKs (or lack thereof) or changing channels (freq) (see https://github.com/lemariva/uPyLoRaWAN/blob/master/sx127x.py for example where the init is a collection of calls to these broken out functions)

This was modelled after how the micropython-lib bluetooth config() function works, with the idea that any keys which aren't set in the argument dict are kept at their current values. So it's possible for one call to either change one or two parameters, or set all the parameters, depending on the arguments.

I believe this makes for smaller code size, but I could have that wrong. @dpgeorge @jimmo is that right, and is this still a good approach?

>    ability to have continuous RX as the default mode as opposed to switching back to standby (that's what I tend to use for powered devices)
>   provide functionality to check for RX in-progress before TX

The first version of this PR had some of this, but I dropped it as I wasn't sure how useful it was, and it adds a bunch of complexity to manage the concurrency around resuming RX cleanly. Will look at adding it back.

My thought was this can be done in this simplified driver by calling receive() in a loop with a relevant timeout, and if also transmitting then call transmit() in between calls to receive() as needed. I was mostly thinking from the point of view of either fairly simple simplex schemes, or LoRaWAN nodes which have specific short receive and transmit windows but the radio is otherwise idle and window timing is controlled by the host.

If you have a different use case with more complex transmit/receive windows and timing that you'd like this to support the please post more details - I'd like to properly understand it.

>    provide a function to calculate packet duration (with LoRa it's not that trivial and often needed since air time can be quite long)
>    provide functions to enter and leave sleep mode, incl. setting all pins to high impedance
 >   provide a function to read the FEI (frequency error indication), I find it pretty useful to tune the frequency of sensor nodes when using FSK mode; I don't know whether LoRa is more robust WRT freq deviation (hmmm, certainly not in the very narrow band modes) and the FEI is an easy way to ensure the nodes agree with a central gateway (I'm not using LoRaWAN)

All good suggestions, will do!

>  I find the way the IRQ flags are managed rather fragile. All the masking/unmasking only works if start_xx/end_xx are
> always correctly called in the right sequence. If they're not, one is guaranteed odd and difficult to troubleshoot results.
> I'm used to clear all flags before switching to TX/RX mode so I'm guaranteed I'm starting from a known state for that
> phase. Also, I've never used the IRQ masks on the semtech chips and rather do a dispatch in the IRQ handler to figure
> out what happened. You seem to experience some spurious interrupt according tot he comments and I wouldn't be
> surprised if it had to do with changing mode, changing IRQ masks, and changing DIO Mapping in back-to-back SPI
> transactions. There's a wimpy little uC in that sx1276 that has to keep up and do the right thing kepping in mind that the > mode change may take way longer than the subsequent SPI transactions (we know that little uC has bugs...)

Goal here was mostly to keep the code small as possible, but this is a reasonable point - especially if it does turn out that masking is causing some of the spurious interrupt issue.

I really only anticipated modem users calling transmit()/receive() directly (sync or async). The start_xx/end_xx /is_receive_finished/is_transmit_finished are primarily an "internal" API for each modem chip. Of course in that case they should be prefixed as such, will change that.

I know there are use cases where the desire is to call start_xyz and end_xyz manually and interleave other computation in between. My feeling was that switching to async and `await m.transmit()` or `await m.receive(timeout)` instead will make the whole thing simpler.

> WRT polling for RX I would not provide any examples or high level code that polls the radio for RX complete. It's a noise 
> disaster in my experience. (Maybe different in stm32wle.) I believe the only appropriate non-interrupt usage pattern for > RX is to have an RX window, start the RX at the beginning of the window, wait, and then check whether something
> arrived once at the end of the window.

Interesting, I agree supporting non-interrupt completion is definitely a sub-optimal fallback. Can you elaborate on "noise disaster"? Do you mean you get significant RF noise injected from the SPI transactions?

> I did not see high-level code to operate the driver in always-RX mode and I believe it's rather tricky. Specifically because
> the IRQ handler gets no info about why it got called and there's an issue with calling RX when no RX-complete indication
> has been provided (e.g. because one wants to TX).

Agree this is not really possible in this version, see the answer above. If you've got some examples of exactly how you'd like to use the driver, that's probably the best way for me to understand what you'd like it to support internally.

> Documentation???

Yes, definitely. This is a Draft PR and there is a TODO checklist in the first post which includes an item for documentation. I have them half-written now but I wanted to get feedback on the API and approach as early as possible. So thanks for providing good feedback on those things here, it might put the actual docs back a little bit though(!) depending if any changes are made.

projectgus commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+
+
+class SX1276(BaseModem):
+    def __init__(self, spi, cs, dio0=None, dio1=None, radio_reset=None, lora_cfg=None):
```
The fields for lora_cfg definitely need to be fully documented!

There can't really be a default because there isn't a single frequency that works everywhere in the world, although that minimal config `{ 'freq_khz': xyz }` should give a working radio in many cases (but recommend to also set `sf`, `bw`, and probably other settings!).

projectgus replied:
Documented now.

projectgus commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+        self._bw_hz = 125000  # Reset value
+        self._t_sym_ms = 1  # Reset value, SF=7 BW=125kHz
+
+        # TODO: decide if it's sensible to put pin config here for board-level SX1276 features like TXCO Enable, RF Antenna switch
```
SX1276 actually has this (an optional `ant_sw` parameter which is an object with `tx()` and `rx()` functions. Need to add the same support here, and then can remove this TODO.

projectgus replied:
Implemented but not yet documented, there's a TODO for that in the first post.

projectgus commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+# LoRa is a registered trademark or service mark of Semtech Corporation or its affiliates.
+
+from micropython import const
+from .modem import BaseModem, ConfigError
```
That's right. I'm not very happy with this hierarchy either, but it seemed like the best of a few poor options.

Could make both `SyncModem()` and `AsyncModem()` into wrappers so you always do something like `SyncModem(SX1726(...))` or `AsyncModem(SX1276(...))`. This does save a very small amount of code size for the "async" user, but it also makes the plain modem classes kind of code-smelly (they're no longer useful for anything on their own).

Plus I think sooner or later there will be functionality that needs to go in a common `BaseModem` class under both SX1276 and SX1262 (maybe the packet time calculation function you requested, for example.)

If you can see a possible hierarchy I've missed, I'm up for thinking about it!

tve replied:
Where do I find `SyncModem`?
I'm not really understanding the issue with `SyncModem(SX1726(...)) `and `AsyncModem(SX1276(...))`.
I don't have a strong opinion here, nor much experience...

projectgus replied:
> Where do I find SyncModem?

No, I just meant this was something I considered as an alternative and then decided it made the "simple case" (no async) unnecessarily fiddly to work with. But I don't have a strong preference either way, so maybe I'll try implementing it.

projectgus replied:
Kept the current hierarchy in the end. It's a little asymmetric but it makes the "Simple API" very simple, which I think is worth the trade-off.

projectgus commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+            pass
+        try:  # output_power
+            pa_config |= (
+                _clamp(lora_cfg["output_power"], 0x0, _PA_CONFIG_OUTPUTPOWER_MASK)
```
yeah, this is very reasonable - will put it on the list to implement.

projectgus replied:
`output_power` is now set in dBm

projectgus commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+        except KeyError:
+            pass
+
+        # Update _REG_PA_CONFIG
```
see general reply about single `config()` with optional keys vs individual functions

projectgus commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+            _REG_IRQ_FLAGS, 0xFF
+        )  # clear all IRQ flags (should be cleared if we have hard reset the radio, but might not be otherwise)
+
+        # Update MODEM_CONFIG1
```
see general reply about single `config()` with optional keys vs individual functions

projectgus commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+        return self._dio0 and (self._dio1 or timeout_ms is None)
+
+    def is_receive_completed(self):
+        # assumption here that _REG_IRQ_FLAGS_MASK is set correctly, so we don't have to check exactly which flags are set
```
yes, this is my mistake I did not notice that the modem will happily deliver you an invalid packet! Thanks

tve replied:
this one caught me too...

projectgus replied:
Driver now drops invalid packets by default, can optionally set a field `rx_crc_error`  to have it return invalid ones (with a flag set on them to indicate invalid CRC).

projectgus commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+        # will_irq if DIO0 and DIO1 both hooked up, or DIO0 and no timeout
+        return self._dio0 and (self._dio1 or timeout_ms is None)
+
+    def is_receive_completed(self):
```
As noted elsewhere, this function shouldn't have  been indicated as part of a public API. The `receive()` function/coro both call it, but they should know the state when they do that.

tve replied:
I do not believe the API in async_modem will be sufficient, i.e., I'm pretty sure I will be calling the various transmit_end/receive_end/etc functions. I provided a good number of examples and maybe you can come up with something that nails it in async_modem...

projectgus replied:
Now there is a new low-level API, and a more complex example (although not as complex as your applications).

projectgus commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+    def end_transmit(self):
+        self._irq_en = ()
+        flags = self._reg_update(_REG_IRQ_FLAGS, _IRQ_TX_DONE, _IRQ_TX_DONE)  # clear flags
+        assert flags & _IRQ_TX_DONE
```
Consistency check that we don't get here unless the `TX_DONE` interrupt flag was set.

Probably makes more sense in light of the other comments about begin_xxx/end_xxx/etc not being intended as public API functions. Sorry about mis-communicating this in the code!

tve replied:
IMHO every assert should at the minimum have a comment that describes what an assertion failure indicates. Specifically here, I don't know whether it would signal invalid use of the end_transmit method by the caller or an unexpected situation WRT the state of the radio.

projectgus commented on micropython/lora/lora/async_modem.py:
```diff
@@ -0,0 +1,42 @@
+        will_cb = self._modem.start_receive(timeout_ms)
+        if will_cb:
+            await self._flag.wait()
+            assert self._modem.is_receive_completed()
```
Should always be the case:

1. start_receive() gets called, sets interrupt mask and clears flags.
2. If `will_cb` is set then we know an interrupt is expected.
3. Interrupt occurs. The only interrupt causes that are masked in should be RX_DONE or RX_TIMEOUT.
4. ISR sets the ThreadSafeFlag, so this task wakes up.
5. Calling `is_receive_completed()` here is just "belts and braces" checking "one of the interrupt flags we masked in is set", to catch the possibility the flag has gotten out of sync with the task.

projectgus replied:
(As noted above though, I missed that with checking on in the modem you can still receive an invalid packet, need to handle this case.)

tve replied:
Would be nice to add the 5 points under "should always be the case" into the source as comment ;-) (or an updated version as you change things)

projectgus replied:
This particular code has changed a lot now, but there are also more comments now.

tve commented:
Re: not breaking config into several smaller functions: I can work with that. Your code does 5-6 unnecessary SPI read-modify-write cycles if given an empty config map, though. That's not so cool. Also, I wonder how long triggering all these try-except blocks takes, plus getting the chip out of sleep mode again.

> If you have a different use case with more complex transmit/receive windows and timing that you'd like this to support the please post more details - I'd like to properly understand it.

Two use-cases:
- I have lots of sensors around the property and have several gateways. The gateways are powered and relay between LoRa and WiFi, they consist of an esp32 + sx1276. The gateways always have RX enabled. When they receive a packet they briefly turn TX on one second later for an ACK. Also, when they receive an outbound packet over WiFi they briefly switch to TX to forward that packet on LoRa.
- I'm just completing a "buddy tracker" that consists of two nodes, each with GPS, and each displaying an arrow pointing in the direction of the other node. Use-case is paddling in the open ocean where it's tough to keep track of where someone else is when the waves are overhead. These nodes TX every second and self-synchronize so they alternate their TX every half second. I'm hoping to extend this to more than two nodes. RX is always on (except for TX) 'cause managing to get an RX is way more important than battery, which has a hefty load due to the display and GPS anyway. In this use-case I could calculate the RX duration such that it ends right when the node needs to TX, but that complicates the app code.

In general, if the power budget is not an issue then why wouldn't you want to have RX always on?

> > I'm used to clear all flags before switching to TX/RX mode so I'm guaranteed I'm starting from a known state for that
> > phase. 
> Goal here was mostly to keep the code small as possible, but this is a reasonable point - especially if it does turn out that masking is causing some of the spurious interrupt issue.

I believe that just writing `0xff` to the IRQ flags and not dealing with the mask at all actually takes less code.

> I know there are use cases where the desire is to call start_xyz and end_xyz manually and interleave other computation in between. My feeling was that switching to async and `await m.transmit()` or `await m.receive(timeout)` instead will make the whole thing simpler.

Please provide an async code example for running in continuous RX with async events causing a need to TX.

> Do you mean you get significant RF noise injected from the SPI transactions?

Yes! I see a big difference in RX sensitivity once you get close to the noise floor if you're constantly polling the IRQ flags. This was a while ago but I know it was on a PCB, i.e., not breadboard with wires.

> > I did not see high-level code to operate the driver in always-RX mode and I believe it's rather tricky. Specifically because
> > the IRQ handler gets no info about why it got called and there's an issue with calling RX when no RX-complete indication
> > has been provided (e.g. because one wants to TX).
> 
> Agree this is not really possible in this version, see the answer above. If you've got some examples of exactly how you'd like to use the driver, that's probably the best way for me to understand what you'd like it to support internally.

I provided some use-cases. For me, this is a complete deal-breaker, i.e., fork&modify situation. Also, I would caution against assuming too many chip-internal states implicitly, like "if this flag is not set it must be that other one and I don't need to check it". In my experience this comes to bite later when misbehavior is discovered.

As an example, I recently saw https://forum.lora-developers.semtech.com/t/sx1262-reduced-rx-sensitivity-packet-reception-fails/162. The failure hypothesis is that AFC/AGC gets off-track due to a strong nearby interference and then the chip misses weak packets that it would normally RX because it's stuck off-frequency and/or low gain. I don't know whether this is what is happening, but I know that a naive driver for sx1231 (rfm69) or sx1276 in FSK mode looks completely different from a good driver, because a good driver has to "watchdog" the chip and reset it if there's any indication that it's stuck. This entails getting interrupts on detecting a high RSSI level, setting a timer, and restarting RX if the chip is still stuck in trying to decode that non-existent packet at the timeout. It has now been so many years that I (and other people) have fought this issue without any ACK from semtech that I wouldn't be surprised if the LoRa chips had that "signature feature" as well. I had hoped to avoid that in LoRa and my heart sank when I bumped into that thread. I haven't confirmed this behavior myself, so I'm secretly hoping there is some other benign issue...

> > Documentation???
> 
> Yes, definitely. This is a Draft PR and there is a TODO checklist in the first post which includes an item for documentation. I have them half-written now but I wanted to get feedback on the API and approach as early as possible. So thanks for providing good feedback on those things here, it might put the actual docs back a little bit though(!) depending if any changes are made.

Ah, I had missed that checkbox, sorry for that. I actually find that writing the docs first is extremely helpful. I often end up with a paragraph explaining something and then realize that I should really change the API so the explanation turns into a short sentence.

Looking forward to the next iteration!

tve commented:
Re: [RX getting stuck](https://forum.lora-developers.semtech.com/t/sx1262-reduced-rx-sensitivity-packet-reception-fails/162) . Notice the sentence: "In the GFSK Rx continuous mode on the SX1268 chip and in the LoRa Rx continuous mode on the SX1276 chip, the problem does not appear." in go.valeriy's post.

Some other choice posts in the forum:
- https://forum.lora-developers.semtech.com/t/sx126x-strange-behavior-after-rx-timeout/1080/3
- https://forum.lora-developers.semtech.com/t/new-sx1276-chip-revision-with-reg-version-0x13/588
- https://forum.lora-developers.semtech.com/t/sx1280-settxparams-and-setsleep/535 (that's sx1280, so not _yet_ relevant, but relevant to the way the config() API works)
- https://forum.lora-developers.semtech.com/t/sx1276-misses-received-lora-packets/436 (no second person confirming the issue)

If you want a headache just browse the forum :-O

The upshot of all this, again, just personal experience, is that as these chip features surface the code gets more complicated and nifty code-saving assumptions no longer hold...

tve commented:
> I really only anticipated modem users calling transmit()/receive() directly (sync or async). The start_xx/end_xx /is_receive_finished/is_transmit_finished are primarily an "internal" API for each modem chip. Of course in that case they should be prefixed as such, will change that.

> I know there are use cases where the desire is to call start_xyz and end_xyz manually and interleave other computation in between. My feeling was that switching to async and await m.transmit() or await m.receive(timeout) instead will make the whole thing simpler.

As you ponder the API, please consider the following, none of which are likely to be satisfied by AsyncModem I assume, perhaps incorrectly:
- I need to get an as precise as possible timestamp for packet RX so I can synchronize nodes, I'm assuming this means I need to write my own interrupt handler, thus I'll be calling the various transmit and receive functions in SX12xx.py directly and will not use Sync/AsyncModem.
- Without using uasyncio for simple nodes I often overlap TX and RX with other stuff. This is often way simpler than throwing asyncio at the problem. Some examples:
 - wake up, get sensor measurement, start TX, go back to sleep (don't bother waiting for TX-done)
 - wake up, get sensor measurement, start TX, low-power-sleep for the expected TX duration, switch radio to sleep, tri-state everything and go to deep sleep
 - sit in a loop consisting of: get sensor values, start TX, do some other work, wait for TX completion, start RX for ACK, do some more other work, wait for RX done/timeout
 - in my tracker the loop is: TX position and wait for completion, start RX, read GPS, update display, sleep, check RX, based on RX timing sleep a small amt more to hit the correct TX timing
- I will want to collect FEI and perhaps packet CRC error stats as well, unless these are exposed this will require dealing with SX12xx.py directly and writing my own interrupt handler
- I will want to periodically query background "RSSI" to get info on the noise floor, again, mess with SX12xx.py directly and write my own high-level loop
- I may finally get around to querying temperature to gauge the need for recalibration and also to have some warning if the chip is overheating (e.g 20dBm+ operation)

tve commented:
Yet something else to consider... In the esp32 lightsleep and deepsleep can be interrupted by a pin level but not edge (IIUC). In order to be woken up from sleep when a packet arrives one would use DIO0 and DIO1 as wake-up pins. In order to be able to go back to sleep after handling the interrupt the pin level has to be low again (again, IIUC), this means the interrupt handler needs to clear the IRQ flag.

NB: this is another use-case for a custom interrupt handler: being able to go back to sleep.

projectgus commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+                << _PA_CONFIG_MAXPOWER_SHIFT
+            )
+            update_mask |= _PA_CONFIG_MAXPOWER_MASK << _PA_CONFIG_MAXPOWER_SHIFT
+        except KeyError:
```
A long time ago I learned from some authoritative-sounding source that "Easier to Ask Forgiveness than Permission" (EAFP) was "more Pythonic", and it stuck. This seems to no longer even be a consensus (was it ever?), I see [Guido pushed back against it in 2014](https://mail.python.org/pipermail/python-dev/2014-March/133118.html).

More importantly we can put aside minor style choices to save code size, and you're right we can save code size here:

```
>>> try:
...     x["a"]
... except KeyError:
...     pass
...     
...     
... 
File <stdin>, code block '<module>' (descriptor: 7fba3cddc160, bytecode @7fba3cddc0c0 31 bytes)
Raw bytecode (code_info_size=5, bytecode_size=26):
 2c 06 01 22 53 48 0c 11 03 11 04 10 02 55 34 01
 59 4a 0a 57 11 05 df 44 43 59 4a 01 5d 51 63
arg names:
(N_STATE 6)
(N_EXC_STACK 1)
  bc=0 line=1
  bc=2 line=2
  bc=21 line=4
00 SETUP_EXCEPT 14
02 LOAD_NAME __repl_print__
04 LOAD_NAME x
06 LOAD_CONST_STRING 'a'
08 LOAD_SUBSCR
09 CALL_FUNCTION n=1 nkw=0
11 POP_TOP
12 POP_EXCEPT_JUMP 24
14 DUP_TOP
15 LOAD_NAME KeyError
17 BINARY_OP 8 
18 POP_JUMP_IF_FALSE 23
20 POP_TOP
21 POP_EXCEPT_JUMP 24
23 END_FINALLY
24 LOAD_CONST_NONE
25 RETURN_VALUE
>>> if "a" in x:
...     x["a"]
...     
...     
... 
File <stdin>, code block '<module>' (descriptor: 7fba3cddc3e0, bytecode @7fba3cddc2e0 23 bytes)
Raw bytecode (code_info_size=4, bytecode_size=19):
 10 04 01 27 10 02 11 03 dd 44 4a 11 04 11 03 10
 02 55 34 01 59 51 63
arg names:
(N_STATE 3)
(N_EXC_STACK 0)
  bc=0 line=1
  bc=7 line=2
00 LOAD_CONST_STRING 'a'
02 LOAD_NAME x
04 BINARY_OP 6 
05 POP_JUMP_IF_FALSE 17
07 LOAD_NAME __repl_print__
09 LOAD_NAME x
11 LOAD_CONST_STRING 'a'
13 LOAD_SUBSCR
14 CALL_FUNCTION n=1 nkw=0
16 POP_TOP
17 LOAD_CONST_NONE
18 RETURN_VALUE
```

projectgus replied:
Switched!

projectgus commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+                else _PA_CONFIG_PASELECT_RFO_PIN
+            )
+            pa_config |= pa_boost_bit
+            update_mask |= pa_boost_bit
```
Yes! Adjusted now.

projectgus commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+            pass
+
+        # RX chain calibration at the specific frequency
+        if rx_calibrate:
```
Added explicit `calibrate()` and `calibrate_image()` functions to the API. Haven't documented yet, have TODO to document.

projectgus commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+        self._irq_callback = callback
+
+    def configure(self, lora_cfg, rx_calibrate=False):
+        self._set_mode(_OPMODE_MODE_SLEEP)
```
Good point, it was not (except for switching in and out of LoRa mode).

projectgus commented on micropython/lora/lora/sx1276.py:
```diff
@@ -0,0 +1,592 @@
+
+            modem_config2 |= sf << _MODEM_CONFIG2_SF_SHIFT
+            update_mask |= _MODEM_CONFIG2_SF_MAX << _MODEM_CONFIG2_SF_SHIFT
+            self._t_sym_ms = 1000 * (1 << sf) // self._bw_hz
```
New function added: `modem.get_time_on_air_us(plen)`, with some tests to verify the times match the Semtech-provided calculator.

projectgus commented:
Big update, lots of additional functionality! @tve @dpgeorge @jimmo Please take a look and let me know what you think.

Still a few small outstanding jobs, but most I've pushed to new PRs as this one is too big already. See first post for details.

The only real downside of adding a lot of functionality is adding code size:

| file | November version (95204e4) | this version (9c90a5995) |
| --- | --- | --- |
| modem.mpy | 539 | 2542 |
| sx126x.mpy | 3599 | 5603 |
| sx127x.mpy | 3320 | 4641 |
| async_modem.mpy | 634 | 1397 |
| Total | 8092 | 14183 |

I feel most of this functionality are things that anyone doing non-trivial LoRa applications will find useful though, and won't be that simple to implement on top of a simpler driver. So I'm not sure I have a better suggestion to make on reducing size here.

projectgus commented on micropython/lora/lora/modem.py:
```diff
@@ -0,0 +1,510 @@
+
+        if not self._rx:
+            # Not actually receiving...
+            return (False, None)
```
@dpgeorge @jimmo In the receive and transmit control flows I now have this pattern of returning 2-tuples and 3-tuples in a few places. In the previous version of this PR these types of results were constrained to "small" types but it got fiddly, not to mention quite confusing to read.

What's your feeling on these additional small heap allocations? I'm thinking I could rework it to allocate a set of list objects and keep them around for reuse, like what's done for the modem command buffers already, but I'm not certain it's worth the complexity vs RAM use tradeoff.

FWIW I think these are the only places that heap allocations can't be avoided during the normal transmit or receive control flow.

projectgus replied:
Following some discussion offline, I've removed these tuples and the normal send/receive code paths are heap allocation free (provided rx_packet instance is passed into receive() and the new packet has the same length.)

projectgus commented:
@tve regarding a couple of these points:

> I will want to collect FEI and perhaps packet CRC error stats as well, unless these are exposed this will require dealing with SX12xx.py directly and writing my own interrupt handler
> I will want to periodically query background "RSSI" to get info on the noise floor, again, mess with SX12xx.py directly and write my own high-level loop
> I may finally get around to querying temperature to gauge the need for recalibration and also to have some warning if the chip is overheating (e.g 20dBm+ operation)

There is some CRC awareness now, the driver by default drops invalid CRC packets and increments a counter `crc_errors` on the modem object that can be read or cleared by the programmer as needed (this needs documenting still). There is also a field option `rx_crc_error` that can be set to True in order for the modem to return messages with invalid CRCs rather than drop them (this part is documented already.)

I want to add an instantaneous RSSI function as well, but this isn't done yet (I'll put it in a future PR, or contribution welcome!).

FEI seems to only be supported on SX1276 for LoRa, which is a bit surprising. On the topic of frequency accuracy, all I see in the SX126x documentation is a guide to keep the bandwidth within the error limit of the clock source and an equation for that. So probably won't make this part of the common modem API, if it's not on any of the newer modems. However, for cases like this then it should always be possible to call `modem._read_reg()` to read out particular register fields and do "advanced" modem-specific stuff. In particular, the packet statistic registers seem to stay valid until the next packet is received so can easily call them after a receive ends and before going back to sleep. Ditto for manually checking the temperature registers and determining if re-calibraiton is needed.

projectgus commented:
>  If you want a headache just browse the forum :-O

Regarding silicon bugs, this is indeed sobering reading.  From what I can see, the only fully confirmed bug in that list is SX126x AGC getting stuck on "min gain" when the receiver gets swamped temporarily. I think probably the best thing to do with these is to document "known" (if not confirmed) silicon bugs in the README and give some workaround tips (like stopping a continuous receive periodically if nothing is received).

I did find what seems like it might be a very nasty silicon bug though, workaround applied in b6eaf74. I still feel like it's too nasty to be that simple, but logic analyzer traces confirm the commands on the wire look correct and the modem otherwise works fine... Might report this one to Semtech as well.

projectgus commented:
Added some more updates and cleanups.

Someone also emailed me to report this driver works well on SX1278 as well (yay!) Looking at the datasheet I don't see any reason not to support all four modems, so it's been updated to do this.

dpgeorge commented on micropython/lora/lora/lora/modem.py:
```diff
@@ -0,0 +1,537 @@
+    # the modem operation is complete, and don't support interleaving transmit
+    # and receive.
+
+    def transmit(self, packet, tx_at_ms=None):
```
IMO (and @jimmo agrees), this driver (and all its components) should use `send` and `recv` naming, like a `socket.socket` does.  (It's not a stream so not read/write.)

That would mean this method is called `send`, and the method below `recv`.

Then it would also have `prepare_send`, `start_send`, `poll_send`, `start_recv`, `poll_recv`, `recv_continuous` etc.

dpgeorge replied:
Some of the reasons for this are:
1. more Pythonic, matches naming elsewhere
2. reduces code size because the names already exist as qstrs

dpgeorge commented on micropython/lora/lora-async/lora/async_modem.py:
```diff
@@ -0,0 +1,149 @@
+        for f in self._flags:
+            f.set()
+
+    # Wrappers for common API calls on the inner modem object
```
Unless the below methods are overridden by the SX12xx class, I think it's possible (and it's simpler, less code) to do:
```py
configure = BaseModem.configure
```

projectgus replied:
Unfortunately these all call through to `self.modem.configure()` not `self.configure`, so I think no easier way to bind it.

(Unless this class was implemented with multiple inheritance instead, see comment below.)

dpgeorge commented:
One thing @jimmo and I just discussed was splitting out the sync-modem methods (namely `BaseModem.transmit`, `BaseModem.receive` and `BaseModem._sync_wait`) into a separate class called `SyncModem`.  That would take a particular SX12xx instance in its constructor, essentially mirroring `AsyncModem`.  It would live in `lora-sync/lora/sync_modem.py`.

Benefits:
- a clear separation of layers (BaseModem -> SX12xx -> SyncModem/AsyncModem)
- no need to have any of the sync modem code if you only use AsyncModem

Drawbacks:
- users must now always install (at least) two packages: one for their hardware, eg `sx126x`, and one for the top-level driver (sync or async)

@projectgus what are your thoughts on that?

projectgus commented:
@dpgeorge 
> One thing @jimmo and I just discussed was splitting out the sync-modem methods (namely `BaseModem.transmit`, `BaseModem.receive` and `BaseModem._sync_wait`) into a separate class called `SyncModem`. That would take a particular SX12xx instance in its constructor, essentially mirroring `AsyncModem`. It would live in `lora-sync/lora/sync_modem.py`.

@tve also asked about the possibility of a change like this, so it's clearly a logical layout for folks.

I agree about the benefits, it's a more symmetrical set of relationships. The reasons I avoided it until now was a general idea of "keep the simple case - Sync - as simple as possible", but also that I don't know what to logically describe these classes if we split them into two types of "thing":

* SyncModem  & AsyncModem are one type of thing (a "modem interface"?)
* SX1276, SX1262, BaseModem, etc are a different type of a thing (a "modem implementation"?)

... not knowing what names to give them makes me think it's a code smell.

(I guess the current code has this problem as well regarding AsyncModem, and maybe if keeping the current structure then AsyncModem should be called AsyncAdapter or something to disambiguate... this is all starting to feel very Java-y!)

The other way around this would be to use multiple inheritance, so each "public" class in the package inherits from either SyncModem or AsyncModem and also from a chip-specific modem base (SX1276, SX1262, etc.), meaning you can do something like `from lora.sync import SX1276`. I generally avoid multiple inheritance (and I'm not even sure how well MP supports it) but actually this might be the most intuitive class structure to go with?

projectgus commented:
Method renames have been implemented!

> One thing @jimmo and I just discussed was splitting out the sync-modem methods (namely BaseModem.transmit, BaseModem.receive and BaseModem._sync_wait) into a separate class called SyncModem. That would take a particular SX12xx instance in its constructor, essentially mirroring AsyncModem. It would live in lora-sync/lora/sync_modem.py.

@dpgeorge @jimmo Please take a look at the head commit in this branch. I'm not 100% sure this is a good idea, but I think it's still preferable to having SyncModem/AsyncModem wrappers that take chip-specific objects as arguments.

I put some reasons in the previous comment on this PR, but the final challenge was telling the user which object to call which method on. For example, do you call `modem.configure(...)` or `modem.modem.configure(...)`? The old AsyncModem wrapper had some hacky workarounds for this like `def configure(self, ...): self.modem.configure(...)`, but this is not a great solution either.

If you think the "mixin-like class" approach in that commit is too fiddly then I fully understand though, I'm not 100% convinced on it myself - the best I could say for it is that it might be the "least bad" way. Am happy to drop thjs commit and try something else instead if you're not convinced.

projectgus commented:
(Rebased, tweaked some wording in the README a tiny bit.)

projectgus commented:
Rebased, pushed a bugfix with CRC as reported by Barry on Discord and another bug fix with async wait timeout on send (only relevant if IRQ doesn't trigger as expected).

dpgeorge commented:
Please make it configure the pins in the constructor (using `hasattr(pin, init)`).  See `sdcard.py` for example.

Also rename u-module to module.

projectgus commented:
@dpgeorge @jimmo Thanks for the review of a large and complex PR! Have applied the requested changes (see [30214ee](https://github.com/micropython/micropython-lib/pull/501/commits/30214eeb78711555e7b8b135a753c42f649f1fd0), [247597b](https://github.com/micropython/micropython-lib/pull/501/commits/247597b95499b93b4c29d26edb49d744f01e348d)), re-tested the two major modem models, squashed commits and rebased on master.

dpgeorge commented:
I tested the `simple_rxtx` code, both sync and async versions, on a pair of SX1276's.  It works!

(I had to fix the `simple_rxtx_async.py` to rename `uasyncio` to `asyncio`, and I pushed that here.)

dpgeorge commented:
Thanks for the huge effort on this @projectgus !

mattytrentini commented:
Congrats and thanks to everyone involved!
